### PR TITLE
Fix handling of `Deref` in `assigning_clones`

### PR DESCRIPTION
**Pull request by Kobzol**
The `assigning_clones` lint had a special case for producing a bit nicer code for mutable references:
```rust
fn clone_function_lhs_mut_ref(mut_thing: &mut HasCloneFrom, ref_thing: &HasCloneFrom) {
    *mut_thing = Clone::clone(ref_thing);
}
//v
fn clone_function_lhs_mut_ref(mut_thing: &mut HasCloneFrom, ref_thing: &HasCloneFrom) {
    Clone::clone_from(mut_thing, ref_thing);
}
```
However, this could [break](https://github.com/rust-lang/rust-clippy/issues/12437) when combined with `Deref`.

This PR removes the special case, so that the generated code should work more generally. Later we can improve the detection of `Deref` and put the special case back in a way that does not break code.

Fixes: https://github.com/rust-lang/rust-clippy/issues/12437

r? @blyxyas

changelog: [`assigning_clones`]: change applicability to `Unspecified` and fix a problem with `Deref`.

### CODE REVIEW

**Comment by blyxyas** [comment]
Hmmmm... What if instead of outright removing it, we could check for the `Deref` lang item?
It would just add a few lines. [Here's the code that you can copy](https://doc.rust-lang.org/clippy/development/trait_checking.html#using-lang-items) and [here's `LangItem::Deref`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/lang_items/enum.LangItem.html)

**Comment by Kobzol** [comment]
I mostly wanted a hotfix so that this doesn't break so often for people, and my experience from writing IDE plugins is that this stuff around `Deref` handling etc. can be incredibly tricky to get right, so it might take some time :)

I will take a look at what you sent me, thanks.

**Comment by Kobzol** [comment]
It's a bit more complicated than I thought. This works fine:
```rust
struct DerefWrapper(DerefInner);

impl Deref for DerefWrapper {
    type Target = DerefInner;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for DerefWrapper {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[derive(Clone)]
struct DerefInner;

fn clone_into_deref_method(mut a: DerefWrapper, b: DerefInner) {
    *a = b.clone();
    a.clone_from(&b);
}
```
But it doesn't work when we uncomment the Clone derive on `DerefWrapper`. So the issue is not `Deref` per-se, but the problem is if both the `Deref` wrapper and the `Deref` target implement `Clone`. Maybe it is unnecessary to also check for this, as just keeping `*` should be enough. I'll try.

**Comment by Kobzol** [comment]
So, some of the stuff works, but I'm having some trouble with recognizing if `Deref` was actually applied. Using this:
```rust
let ty = cx.typeck_results().expr_ty(ref_expr);
let impls_deref = cx.tcx.lang_items().deref_mut_trait().map_or(false, |id| implements_trait(cx, ty, id, &[]));
```
also fires here:
```rust
fn clone_function_lhs_mut_ref(mut_thing: &mut HasCloneFrom, ref_thing: &HasCloneFrom) {
    *mut_thing = Clone::clone(ref_thing);
}
```
because the LHS has type `&ReErased mut HasCloneFrom`, which apparently implements `Deref`, even though `HasCloneFrom` doesn't.

I tried using `typeck_results().expr_adjustments()`, but that doesn't actually contain a `Deref` adjustment even if LHS goes through a `Deref` type...

**Comment by Kobzol** [comment]
I pushed my code. The second commit shows that some unnecessary `&mut *` is being generated.

**Comment by gibfahn** [comment]
>Hmmmm... What if instead of outright removing it, we could check for the Deref lang item?

Could we remove the special-case first, and then continue this PR to add it back later? Having clippy autofix working code to broken code (even on nightly) feels like a "revert first, improve later" situation to me (as a user).

**Comment by Kobzol** [comment]
> Could we remove the special-case first, and then continue this PR to add it back later?

I agree, that's why I changed the applicability [here](https://github.com/rust-lang/rust-clippy/pull/12554). Therefore this lint should no longer autofix.

**Comment by xFrednet** [comment]
Hey @Kobzol, this is a ping from triage, since there hasn't been any activity in some time. Are you still planning to continue this implementation?

If you have any questions, you're always welcome to ask them in this PR or on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy).

@rustbot author

**Comment by Kobzol** [comment]
Well, it was waiting for a review :) I'll try to rebase it.

**Comment by xFrednet** [comment]
Ahh okay, I guessed that but wasn't sure. Then let's pick a new reviewer:

r? clippy

**Comment by xFrednet** [comment]
r? clippy

**Comment by xFrednet** [comment]
Hey @Alexendoo, rustbot has picked you as a new reviewer. If you don't have the time, you can give the PR to me. Just comment `r? @xFrednet`

**Comment by Kobzol** [comment]
Rebased. It's still WIP, but I'd appreciate some feedback on the approach that I have used here.

File: clippy_lints/src/assigning_clones.rs
Comment (Alexendoo) [inline clippy_lints/src/assigning_clones.rs:383]: How about only checking if `ty.is_ref()`? Same for below. They're the only safe types deref wise, it would add an explicit deref for `!Clone` wrappers but I think that's good as it avoids any future issues if the wrapper becomes `Clone`

It would also handle e.g. `*mut Waker`

Comment (Kobzol) [reply clippy_lints/src/assigning_clones.rs]: Sorry for the delay, was on vacation season. Good idea, this is probably the most conservative solution. Pushed a commit that does this

Comment (Alexendoo) [reply clippy_lints/src/assigning_clones.rs]: Looks good thanks, just needs the tests blessing

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
Great, thank you!

@bors r+

File: clippy_lints/src/assigning_clones.rs
Comment (Kobzol) [reply clippy_lints/src/assigning_clones.rs]: Done :)

**Pull request merged**
